### DISCUSSION TITLE
Valetudo 2021.12.0

### DISCUSSION THREAD

**Discussion by Hypfer** [Hypfer/Valetudo#1259] [Release Discussions]
<div align="center">
    <img src="https://raw.githubusercontent.com/Hypfer/Valetudo/master/assets/logo/valetudo_logo_with_name.svg" width="800" alt="valetudo">
    <p align="center"><h2>2021.12.0</h2></p>
</div>

Map data format changes, quality of life improvements and bugfixes.

## Map data format changes

This release features the first version increment of the ValetudoMap data format.
Initially, in version 1, pixel data such as walls or floor were stored as a huge array of coordinate pairs: 
`[x0, y0, x1, y1, ...]`. This was not a very efficient way of storing that data as most of it is redundant.

This resulted in map files that were over 1 MiB in size being pushed to your browser every 3 seconds.

To combat this, some thought was put into how the data is stored.
If you have a row of pixels, all of them will have the same y-coordinate, meaning that you only have to store that once.
People already figured that out [in the 60s](https://en.wikipedia.org/wiki/Run-length_encoding).

With Map V2, pixels are stored as `[xStart0, y0, count0, xStart1, y1, count1, ...]` where `[xStart, y]` is the starting point and `count` represents additional pixels with the same y-coordinate to the right.

This resulted in V2 map files using between 5% and 15% of the size of the equivalent V1 map.
The huge testfile map for example was compressed from 847 kB to 75 kB.

While this is a **breaking change**, the breakage actually isn't that big of a change as V2 can easily be converted back to V1.
In fact, running `JSON.parse()` on a V2 map and then converting it to a V1 one in memory is more than twice as fast as running `JSON.parse()` on a V1 map.

Since this change drastically reduces the traffic caused by live maps, I expect to see general usability and reliability
improvements in situations where the Wi-Fi reception isn't that great.

## LogViewer

The LogViewer has been greatly improved and now features a structured and color-coded view of your log.

[<img src="https://user-images.githubusercontent.com/974410/144744519-5ba963b9-6409-4e04-95d9-e73caf55021c.png" width=700>](https://user-images.githubusercontent.com/974410/144744519-5ba963b9-6409-4e04-95d9-e73caf55021c.png)

You can use the text filter at the top to either filter for the loglevel or the message contents.
Just start typing.

## UI Help

Both consumables and timers now feature a help button, which shall explain how to use these features.

[<img src="https://user-images.githubusercontent.com/974410/144744625-d113d697-e4e1-4f77-a3a2-596ad88131a5.png" width=700>](https://user-images.githubusercontent.com/974410/144744625-d113d697-e4e1-4f77-a3a2-596ad88131a5.png)

## Misc

- Map reset on newer roborock models should now be fixed thanks to @supersmile2009
- Zooming the map should now work much better if you zoom via a touchpad or trackpoint instead of a mousewheel
- MQTT current statistics now have a unit_of_measurement in HA
- Lowmem builds now have a reduced map poll rate to reduce memory pressure
- Valetudo now manages its logfile size to prevent it from growing indefinitely

### Features

* **logger:** Automatically clear the logfile when it gets too large ([c0698b4](https://github.com/Hypfer/Valetudo/commit/c0698b43dda5144d7d0c2c58d9336a1c0662cc65))
* Reduce map polling interval for lowmem hosts ([fcf94ce](https://github.com/Hypfer/Valetudo/commit/fcf94ceeb60f9b46e25bfa04d2e4e88ac2a25926))
* **ntpClient:** Improve log messages ([8880bbf](https://github.com/Hypfer/Valetudo/commit/8880bbfc7ed79969a61c6d6e51f7023c23d7561e))
* **ui:** Add help button + modal to consumables view ([d1388f1](https://github.com/Hypfer/Valetudo/commit/d1388f124dae0046d2bee49345d0513441926f21))
* **ui:** Add timers help ([c864170](https://github.com/Hypfer/Valetudo/commit/c864170d44c2b686394bb6d770cff937e8a9de43))
* **ui:** Implement LogViewer component ([1b96ae0](https://github.com/Hypfer/Valetudo/commit/1b96ae06c8a74acbc9a6fa03cf941736520e30d0))
* **ui:** Remove obsolete about > github release info ([253d9f9](https://github.com/Hypfer/Valetudo/commit/253d9f99b0965859705c52ffae9483c381d49ad7))
* **ui:** Swap Manual control cw & ccw rotation button ([44eab06](https://github.com/Hypfer/Valetudo/commit/44eab06eadcd57c8c9b3784bc6f4c7a6dbea58e8))

### Bug Fixes

* **ui:** Try avoiding bright flashes on initial loading ([f84a51d](https://github.com/Hypfer/Valetudo/commit/f84a51dc7c09ec4a7cb19e40e3889544375af58a))
* Fix log sse endpoint ([fd6f0e2](https://github.com/Hypfer/Valetudo/commit/fd6f0e2d34ebc2f9de4aa5b23da2fa8ec2e4d630)), closes [#1256](https://github.com/Hypfer/Valetudo/issues/1256)
* **mqtt:** Fix unit_of_measurement for current statistics and consumables ([7f052dc](https://github.com/Hypfer/Valetudo/commit/7f052dc3f64905afd69e35de217cd390377e6ce2))
* **ntpClient:** Delay initial ntp sync to wait for valetudo to start up properly ([727af04](https://github.com/Hypfer/Valetudo/commit/727af04ab54d007c21a3c6ac270036611280e112))
* **ui:** AppBar and menu should not be user-selectable ([69754b7](https://github.com/Hypfer/Valetudo/commit/69754b7204efe81ba147539a04630d644d0e07f4))
* **ui:** Cleanup unused code ([c050940](https://github.com/Hypfer/Valetudo/commit/c05094011540c46737e7829532a319756255b2f3))
* **ui:** Fix map zoom behavior for trackpoint and trackpad scroll zoom ([ca94600](https://github.com/Hypfer/Valetudo/commit/ca9460095eb6e4946f7653328667e8df60c678dc))
* **ui:** Updater changelog should stay in its constraints ([5d815bf](https://github.com/Hypfer/Valetudo/commit/5d815bf98f334b9f40faecaf00f918bb6fecaacb))
* **updater:** Ensure that there is enough space besides process.argv0 to store the new binary ([c298e98](https://github.com/Hypfer/Valetudo/commit/c298e9833cc6dcb2457a52b986daa8f2e81a8850))
* **vendor.dreame:** Fix wrong segment iteration count ([9fd9c17](https://github.com/Hypfer/Valetudo/commit/9fd9c1723e0329b76b89e84cca108cc117a2a736))
* **vendor.dreame:** There are actually no firmware limits for virtual restrictions ([8a8b536](https://github.com/Hypfer/Valetudo/commit/8a8b536b2eb126b9604b58e58cf196824da63be1))
* **vendor.roborock:** Fix resetting incomplete map on Roborock. Detect and reset current map slot. ([#1252](https://github.com/Hypfer/Valetudo/issues/1252)) ([ba86057](https://github.com/Hypfer/Valetudo/commit/ba86057ba7946ee90c6ee7408ca446de56d23136))
* Fix CurrentStatistics UI refresh ([93249ad](https://github.com/Hypfer/Valetudo/commit/93249ad142d4879ea967898f26905645af1c4f2b))
* **ui:** Fix LogViewer overflow-wrap ([4799027](https://github.com/Hypfer/Valetudo/commit/4799027a42e6f5216cb748c75b88b1291a863335))

<hr /><em>This discussion was created from the release <a href='https://github.com/Hypfer/Valetudo/releases/tag/2021.12.0'>Valetudo 2021.12.0</a>.</em>

**Comment by Yax** [comment]
`valetudo_release_manifest.json` might be missing in the artifacts.

**Reply by Hypfer** [reply]
Every release gets built by a github action which takes around 15 minutes to complete after the release has been published.

You're just too fast.

**Reply by Yax** [reply]
Oh, thanks for the info! I didn't know that. Anyway - thanks for the update!
